p2pserver/protocols/subnet/subnet.go

```diff
@@ -0,0 +1,383 @@
+}
+
+func (self *SubNet) getUnconnectedGovNode() []string {
+	self.lock.Lock()
+	defer self.lock.Unlock()
+
+	var addrs []string
+	for addr := range self.members {
+		if self.connected[addr] == nil {
+			addrs = append(addrs, addr)
+		}
+	}
+
+	return addrs
+}
+
+func (self *SubNet) newMembersRequest(from, to common.PeerId) *types.SubnetMembersRequest {
+	var request *types.SubnetMembersRequest
+	if self.IsSeedNode() {
+		request = types.NewMembersRequestFromSeed()
+	} else if self.acct != nil && self.gov.IsGovNodePubKey(self.acct.PublicKey) {
+		var err error
+		request, err = types.NewMembersRequest(from, to, self.acct)
+		if err != nil {
+			return nil
+		}
+	}
+
+	return request
+}
+
+func (self *SubNet) sendMembersRequestToRandNodes(net p2p.P2P) {
+	count := 0
+	peerIds := make([]common.PeerId, 0, MaxMemberRequests)
+	self.lock.RLock()
+	// note map iteration is randomized
+	for _, p := range self.connected {
+		count += 1
+		peerIds = append(peerIds, p.Id)
+		if count == MaxMemberRequests {
+			break
+		}
+	}
+	self.lock.RUnlock()
+
+	for _, peerId := range peerIds {
+		request := self.newMembersRequest(net.GetID(), peerId)
+		if request == nil {
+			return
+		}
+		net.SendTo(peerId, request)
+	}
+}
+
+func (self *SubNet) sendMembersRequest(net p2p.P2P, peer common.PeerId) {
+	request := self.newMembersRequest(net.GetID(), peer)
+	if request == nil {
+		return
+	}
+
+	net.SendTo(peer, request)
+}
+
+func (self *SubNet) cleanInactiveGovNode() {
+	now := time.Now()
+	self.lock.Lock()
+	defer self.lock.Unlock()
+
+	for addr, member := range self.members {
+		if member.Alive.Add(MaxInactiveTime).Before(now) {
+			delete(self.members, addr)
+		}
+	}
+}
+
+func (self *SubNet) cleanRetiredGovNode(net p2p.P2P) {
+	self.lock.Lock()
+	defer self.lock.Unlock()
+	for addr, status := range self.members {
+		if !self.gov.IsGovNode(status.PubKey) {
+			delete(self.members, addr)
+			if info := self.connected[addr]; info != nil {
+				if p := net.GetPeer(info.Id); p != nil {
+					p.Close()
+				}
+			}
+		}
+	}
+}
+
+func (self *SubNet) maintainLoop(net p2p.P2P) {
+	parker := self.unparker
+	for {
+		self.lock.Lock()
+		for _, p := range net.GetNeighbors() {
+			listen := p.Info.RemoteListenAddress()
+			if self.members[listen] != nil && self.connected[listen] == nil {
+				self.connected[listen] = p.Info
+				self.members[listen].Alive = time.Now()
+			}
+		}
+		seedOrGov := self.IsSeedNode() || (self.acct != nil && self.gov.IsGovNodePubKey(self.acct.PublicKey))
+		self.lock.Unlock()
+
+		self.cleanRetiredGovNode(net)
+		for _, addr := range self.getUnconnectedGovNode() {
+			self.logger.Infof("[subnet] try connect gov node: %s", addr)
+			go net.Connect(addr)
+		}
+
+		self.cleanInactiveGovNode()
+
+		if seedOrGov {
+			self.sendMembersRequestToRandNodes(net)
+			members := self.GetMembersInfo()
+			buf, _ := json.Marshal(members)
+			self.logger.Infof("[subnet] current members: %s", string(buf))
+		}
+
+		parker.ParkTimeout(RefreshDuration)
```

Comment:
maintainloop seems not an idle loop.  
It will be busy with sending members-req & unparked by members-response.

Reply:
The loop should be idle if there's no change on members.

Reply:
updated